### PR TITLE
Fix issue 21177 - printf check does not recognize GNU extensions

### PR DESCRIPTION
**Pull request by Luhrel**

### CODE REVIEW

File: src/dmd/chkformat.d [hunk 1]
```diff
@@ -915,6 +940,26 @@ Format parseGenericFormatSpecifier(scope const char[] format,
             break;
     }
 
+    version (CRuntime_Glibc)
```
Comment (MoonlightSentinel) [inline src/dmd/chkformat.d:943]: This depends on the host compiler - not the currently targeted `libc`

Comment (Luhrel) [reply src/dmd/chkformat.d:943]: Is there an easy way to enable this test only for the glibc runtime ?

Comment (Luhrel) [reply src/dmd/chkformat.d:943]: No, it's not. DMD can use the GNU C library, just like GDC and LDC.

Comment (MoonlightSentinel) [reply src/dmd/chkformat.d:943]: Not sure what you are talking about, `version (CRuntime_Glibc)` is evaluated when *compiling dmd itself* (regardless of the actual host compiler).
But these extensions are only available when the *user code* is linked to glibc.

Comment (Luhrel) [reply src/dmd/chkformat.d:943]: You're right 🤦🏼

**Review by ibuclaw — CHANGES_REQUESTED** [review]

File: src/dmd/chkformat.d [hunk 2]
```diff
@@ -260,6 +276,9 @@ bool checkPrintfFormat(ref const Loc loc, scope const char[] format, scope Expre
                 deprecation(loc, "format specifier `\"%.*s\"` is invalid", cast(int)slice.length, slice.ptr);
                 break;
 
+            case Format.m:
+                break;  // not assert(0) because it may go thought it if there is more arguments than format specifiers
```
[the remark this answers is not in the document]
Comment (Luhrel) [reply src/dmd/chkformat.d:280]: Yes, bad autocorrect.

File: src/dmd/chkformat.d [hunk 3]
```diff
@@ -915,6 +940,26 @@ Format parseGenericFormatSpecifier(scope const char[] format,
             break;
     }
 
+    version (CRuntime_Glibc)
+    {
+        // GNU extentions
```
[the remark this answers is not in the document]
Comment (Luhrel) [reply src/dmd/chkformat.d:945]: it's a typo

File: src/dmd/cond.d [hunk 4]
```diff
@@ -961,6 +961,7 @@ extern (C++) final class StaticIfCondition : Condition
  * Returns:
  *      true if found
  */
+@safe nothrow pure
```
Comment (Geod24) [inline src/dmd/cond.d:964]: Those attributes are usually put after the function, on the RHS

Comment (Geod24) [reply src/dmd/cond.d:964]: And I see the `chkformat` module does it differently...

File: src/dmd/chkformat.d [hunk 5]
```diff
@@ -915,6 +942,26 @@ Format parseGenericFormatSpecifier(scope const char[] format,
...
     }
 
+    if (findCondition(global.versionids, Identifier.idPool("CRuntime_Glibc")))
+    {
+        // GNU extensions
+        // printf: http://www.gnu.org/software/libc/manual/html_node/Other-Output-Conversions.html
+        // scanf: https://www.gnu.org/software/libc/manual/html_node/Dynamic-String-Input.html
+        switch (sc)
+        {
+            case 'a':
```
Comment (Geod24) [inline src/dmd/chkformat.d:952]: It seems that `a` is already handled up there ? Maybe just merge the switches and have a `findCondition` in there.

File: src/dmd/chkformat.d [hunk 6]
```diff
@@ -493,7 +518,7 @@ private:
  * Returns:
  *      Format
  */
-pure nothrow @safe
+nothrow @safe
```
Comment (Geod24) [inline src/dmd/chkformat.d:521]: You could keep the attribute by passing `bool useGNUExts = findCondition(...)` as a default argument.

File: test/compilable/chkformat.d [hunk 7]
```diff
@@ -1,9 +1,12 @@
 // https://issues.dlang.org/show_bug.cgi?id=20643
 // https://issues.dlang.org/show_bug.cgi?id=20644
+// https://issues.dlang.org/show_bug.cgi?id=21177
 /*
+CC: g++
```
Comment (Geod24) [inline test/compilable/chkformat.d:5]: Why is this necessary ?

**Review by Geod24 — COMMENTED** [review]
Thanks! Could you also rename the commit to follow our standards, that is, include the title of the issue as well, not just "Fix issue XXXX" which is very cryptic when one looks at the history ?

File: test/compilable/chkformat.d (hunk 7 above)
Comment (Luhrel) [reply test/compilable/chkformat.d:5]: I thought that would limit the test for the Glibc runtime, but apparently not. Is there an easy way to achieve this ? (Is glibc limited to Linux ?) If not I may use dshell.

Comment (Geod24) [reply test/compilable/chkformat.d:5]: See `DISABLED` [here](https://github.com/dlang/dmd/tree/master/test#test-configuration).
But you will need to create another test, as we don't want to disable `chkformat.d` test on Windows...

Comment (MoonlightSentinel) [reply test/compilable/chkformat.d:5]: No, that's not possible right now.
Guess you can either use dhsell test or try to fake the error messages without glibc using `pragma(msg, ...)`

File: src/dmd/chkformat.d (hunk 6 above)
Comment (Luhrel) [reply src/dmd/chkformat.d:521]: Nice trick !

File: test/compilable/chkformat.d (hunk 7 above)
Comment (Luhrel) [reply test/compilable/chkformat.d:5]: > try to fake the error messages without glibc using `pragma(msg, ...)`

Nice trick !

**Comment by Luhrel** [comment]
> The way gcc does this is by providing two different attributes: printf and gnu_printf. So the equivalent here is to introduce pragma(gnu_printf) and pragma(gnu_scanf), then update the druntime bindings for CRuntime_Glibc declarations accordingly.

@ibuclaw doesn't printf implicitly calls gnu_printf under Glibc ?

**Comment by Luhrel** [comment]
Apparently Windows doesn't like the test 21177 🤔

**Comment by ibuclaw** [comment]
> 
> @ibuclaw doesn't printf implicitly calls gnu_printf under Glibc ?

Maybe, it seems reasonable that would be the default assumption if the target doesn't provide its own format types.

I've noticed that there's also `ms_printf`, and Darwin and Solaris have their own format specifiers as well.

**Comment by Luhrel** [comment]
> 
> 
> > @ibuclaw doesn't printf implicitly calls gnu_printf under Glibc ?
> 
> Maybe, it seems reasonable that would be the default assumption if the target doesn't provide its own format types (just mentioning an internal implementation detail).
> 
> I've noticed that there's also `ms_printf`, and Darwin and Solaris have their own format specifiers as well.

I will implement them in a different PR.

**Comment by ibuclaw** [comment]
> > @ibuclaw doesn't printf implicitly calls gnu_printf under Glibc ?
> 
> Maybe, it seems reasonable that would be the default assumption if the target doesn't provide its own format types (just mentioning an internal implementation detail).

According to [gcc's](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-format-function-attribute) common function attribute documentation:

> _archetype_ values such as `printf` refer to the formats accepted by the system’s C runtime library, while values prefixed with `gnu_` always refer to the formats accepted by the GNU C Library. On Microsoft Windows targets, values prefixed with `ms_` refer to the formats accepted by the msvcrt.dll library.

Though I guess we can leave the jury to decide on whether the same should be imitated here.

**Comment by RazvanN7** [comment]
@ibuclaw It seems like your comments have been addressed, is there anything else that you require from this PR or can we merge it?

File: src/dmd/chkformat.d [hunk 8]
```diff
@@ -910,6 +943,15 @@ Format parseGenericFormatSpecifier(scope const char[] format,
...
             break;
 
+        case 'm':
+            if (useGNUExts)
+            {
+                // http://www.gnu.org/software/libc/manual/html_node/Other-Output-Conversions.html
+                specifier = Format.GNU_m;
+                break;
+            }
+            goto default;
```
Comment (RazvanN7) [inline src/dmd/chkformat.d:953]: This line is not covered by tests.

File: src/dmd/chkformat.d [hunk 9]
```diff
@@ -439,15 +460,19 @@ bool checkScanfFormat(ref const Loc loc, scope const char[] format, scope Expres
...
                     errorMsg(null, slice, e, "double*", t);
                 break;
+
             case Format.Lg:     // pointer to long double
                 if (!(t.ty == Tpointer && tnext.ty == Tfloat80))
                     errorMsg(null, slice, e, "real*", t);
                 break;
 
+            case Format.GNU_a:
+            case Format.GNU_m:
```
Comment (RazvanN7) [inline src/dmd/chkformat.d:475]: These 2 lines are not covered by tests.

File: src/dmd/chkformat.d [hunk 10]
```diff
@@ -113,15 +124,21 @@ bool checkPrintfFormat(ref const Loc loc, scope const char[] format, scope Expre
 
         if (precisionStar)
         {
-            auto e = getNextArg();
+            bool skip;
+            auto e = getNextArg(skip);
+            if (skip)
+                continue;
```
Comment (RazvanN7) [inline src/dmd/chkformat.d:130]: This line is not covered by tests.

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
All added lines should be covered by tests.

File: test/compilable/test21177.d [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+---
+compilable/test21177.d(19): Deprecation: more format specifiers than 0 arguments
+---
+*/
+
+import core.stdc.stdio;
+
+void main()
+{
+    version (CRuntime_Glibc)
```
Comment (ibuclaw) [inline test/compilable/test21177.d:14]: You might want to add a `#line` to keep the test output stable for any future additions.

**Review by RazvanN7 — APPROVED** [review]

File: src/dmd/chkformat.d [hunk 12]
```diff
@@ -439,15 +460,19 @@ bool checkScanfFormat(ref const Loc loc, scope const char[] format, scope Expres
...
                 if (!(t.ty == Tpointer && tnext.ty == Tfloat64))
                     errorMsg(null, slice, e, "double*", t);
                 break;
+
             case Format.Lg:     // pointer to long double
                 if (!(t.ty == Tpointer && tnext.ty == Tfloat80))
                     errorMsg(null, slice, e, "real*", t);
                 break;
 
+            case Format.GNU_a:
```
Comment (ibuclaw) [inline src/dmd/chkformat.d:474]: Wrong, the argument to `%a` must be `char**`, not `char*`.

File: src/dmd/chkformat.d (hunk 9 above)
Comment (ibuclaw) [inline src/dmd/chkformat.d:475]: This is also not correct.
